gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/webauthn/WebAuthnRegisterHandler.java

```diff
@@ -0,0 +1,252 @@
+        final Client client = ctx.get(ConstantKeys.CLIENT_CONTEXT_KEY);
+        final String username = session.get(ConstantKeys.PASSWORDLESS_CHALLENGE_USERNAME_KEY);
+        final String credentialId = webauthnResp.getString("id");
+
+        // authenticate the user
+        webAuthn.authenticate(
+                // authInfo
+                new WebAuthnCredentials()
+                        .setOrigin(origin)
+                        .setChallenge(session.get(ConstantKeys.PASSWORDLESS_CHALLENGE_KEY))
+                        .setUsername(session.get(ConstantKeys.PASSWORDLESS_CHALLENGE_USERNAME_KEY))
+                        .setWebauthn(webauthnResp), authenticate -> {
+
+                    // invalidate the challenge
+                    session.remove(ConstantKeys.PASSWORDLESS_CHALLENGE_KEY);
+                    session.remove(ConstantKeys.PASSWORDLESS_CHALLENGE_USERNAME_KEY);
+
+                    if (authenticate.succeeded()) {
+                        // create the authentication context
+                        final AuthenticationContext authenticationContext = createAuthenticationContext(ctx);
+                        // update the credential
+                        updateCredential(authenticationContext, credentialId, authenticatedUser.getId(), credentialHandler -> {
+                            if (credentialHandler.failed()) {
+                                logger.error("An error has occurred while updating credential for user {}", username, credentialHandler.cause());
+                                ctx.fail(401);
+                                return;
+                            }
+                            if (isEnrollingFido2Factor(ctx)) {
+                                enrollFido2Factor(ctx, authenticatedUser, createEnrolledFactor(session.get(ENROLLED_FACTOR_ID_KEY), credentialId));
+                            } else {
+                                manageFido2FactorEnrollment(ctx, client, credentialId, authenticatedUser);
+                            }
+                            ctx.next();
+                        });
+                    } else {
+                        logger.error("Unexpected exception", authenticate.cause());
+                        ctx.fail(authenticate.cause());
+                    }
+                });
+    }
+
+    private AuthenticationContext createAuthenticationContext(RoutingContext context) {
+        HttpServerRequest httpServerRequest = context.request();
+        SimpleAuthenticationContext authenticationContext = new SimpleAuthenticationContext(new VertxHttpServerRequest(httpServerRequest.getDelegate()));
+        authenticationContext.set(Claims.ip_address, RequestUtils.remoteAddress(httpServerRequest));
+        authenticationContext.set(Claims.user_agent, RequestUtils.userAgent(httpServerRequest));
+        authenticationContext.set(Claims.domain, domain.getId());
+        return authenticationContext;
```

Comment:
this method is duplicate and can be moved into the WebAuthnHandler

Reply:
FP